Add payload size metric

lmsilva-wls opened the pull request.

lmsilva-wls committed:
Add payload size metric

mahtues commented on metrics/metrics.go:
```diff
@@ -76,7 +88,7 @@ var (
 			Name:      "response_time_ms",
 			Help:      "the response time in ms of calls to server",
-			Buckets:   []float64{1, 3, 5, 10, 25, 50, 100, 150, 200, 250, 300},
+			Buckets:   defaultLatencyBuckets(),
```
👍

mahtues commented on app/app.go:
```diff
@@ -163,17 +164,27 @@ func (a *App) metricsReporterInterceptor(
 
+	// using JSON to calculate the payload size introduces a small, invariant/constant error related when compared to
+	// the gRPC payload, but should be enough for observability
+	marshalledEvents, _ := json.Marshal(events)
```
`protobuf` compacts field names. Serializing it into JSON will add field names. The JSON size will be greater than the wire size.

mahtues replied:
![image](https://github.com/topfreegames/eventsgateway/assets/14203456/cf66a198-5f16-491b-b986-7e907aeacb3b)

![image](https://github.com/topfreegames/eventsgateway/assets/14203456/123d4f29-fb1a-4edb-8ca6-b6e0ffb28967)

mahtues replied:
on second tought, calculate the payload inside the type switch.
![image](https://github.com/topfreegames/eventsgateway/assets/14203456/581c449c-e85f-4b4d-adfc-965d69bc96b1)

lmsilva-wls replied:
Nice catch! I've updated it

mahtues replied:
👍

mahtues commented on app/app.go:
```diff
@@ -24,6 +24,7 @@ package app
 
 import (
 	"context"
+	"encoding/json"
```
```suggestion
	"github.com/golang/protobuf/proto"
```

mahtues commented on app/app.go:
```diff
@@ -163,17 +164,27 @@ func (a *App) metricsReporterInterceptor(
 		WithField("route", info.FullMethod).
 		WithField("topic", topic)
 
+	// using JSON to calculate the payload size introduces a small, invariant/constant error related when compared to
+	// the gRPC payload, but should be enough for observability
+	marshalledEvents, _ := json.Marshal(events)
+	payloadSize := len(marshalledEvents)
```
```suggestion
	payloadSize := 0
	for _, e := range events {
		payloadSize += proto.Size(e)
	}
```

lmsilva-wls committed:
Decrease bucket sizes

lmsilva-wls committed:
Add configuration overrides for buckets on histogram metrics

lmsilva-wls committed:
Add client side metrics to the default registry

With this, every application must expose their own prometheus HTTP handler for the
metrics endpoint; after that is done, the EG client metrics will be available on the same
registry.

lmsilva-wls committed:
Add docker compose setup for prometheus and grafana
